guardian/cdk
# feat: Add pattern for Lambda which processes Kinesis Stream

## What does this change?

This PR adds a pattern for lambdas which process Kinesis streams.

Example usage:

```typescript
new GuKinesisLambda(stack, "my-lambda-function", {
  code: { bucket: "test-dist", key: "lambda.zip" },
  functionName: "my-lambda-function",
  handler: "my-lambda/handler",
  runtime: Runtime.NODEJS_14_X,
  errorHandlingConfiguration: {
    bisectBatchOnError: true,
    retryBehaviour: StreamRetry.maxAttempts(3),
  },
  monitoringConfiguration: {
    toleratedErrorPercentage: 1,
    snsTopicName: "alerts-topic",
  },
});
```

## How to test

1. Create a new stack based on this pattern. You should end up with a functional lambda triggered whenever new records are put onto a Kinesis stream.
1. Migrate existing stacks to use this pattern:
    1. A stack which defines its own Kinesis stream
    1. A stack which uses an external Kinesis stream

## How can we measure success?

* It should be faster to set up new projects correctly using this pattern
* When migrating existing projects, users can configure sensible monitoring for a tiny amount of extra code/effort
* When migrating existing projects, users will be encouraged to configure appropriate error handling. Currently this is optional, which means that we often use [AWS's (potentially unhelpful) defaults](https://aws.amazon.com/blogs/compute/new-aws-lambda-controls-for-stream-processing-and-asynchronous-invocations/).

## Have we considered potential risks?

It's tricky to test the migration scenarios (outside of unit tests) as Developer Experience don't seem to have any stacks which use this architecture. We will need to find some volunteers to work on trying this pattern out with us.
